### PR TITLE
RavenDB-13106 Support subscriptions with resharding

### PR DESCRIPTION
**Pull request by karmeli87**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18177

### Additional description

This PR ensures that the subscriptions guarantees are maintained when the resharding is triggered.

Most of the changes here are affecting only the sharded subscriptions expect for one important change:
Now we _keep the entire batch of documents in memory_ before we send it to the client, because it is required to go thtough the state machine first and drop from the batch documents that has been already processed.
This change affects also regular databases because we didn't want to have a separate logic for that.

Missing in this PR:
- Handle sharded revisions subscriptions while resharding (block entirly or implement)
- Handle failover and resharding (failover for sharded database is working)

### Type of change

- New feature

### How risky is the change?

- High

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- Yes. Please list the affected platforms.
- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Client/Documents/Operations/OngoingTasks/GetTaskInfoResult.cs [hunk 1]
```diff
@@ -94,8 +94,6 @@ public override DynamicJsonValue ToJson()
             json[nameof(Query)] = Query;
             json[nameof(SubscriptionName)] = SubscriptionName;
             json[nameof(SubscriptionId)] = SubscriptionId;
-            json[nameof(MentorNode)] = MentorNode;
-            json[nameof(PinToMentorNode)] = PinToMentorNode;
```
Comment (ayende) [inline src/Raven.Client/Documents/Operations/OngoingTasks/GetTaskInfoResult.cs:98]: We don't have that for subscriptions any longer? Why?

File: src/Raven.Client/Documents/Subscriptions/SubscriptionState.cs [hunk 2]
```diff
@@ -17,17 +15,20 @@ public class SubscriptionState : IDatabaseTask, IDatabaseTaskStatus
...
         public string MentorNode { get; set; }
         public bool PinToMentorNode { get; set; }
-        public string NodeTag { get; set; }
+
         public DateTime? LastBatchAckTime { get; set; }  // Last time server made some progress with the subscriptions docs  
         public DateTime? LastClientConnectionTime { get; set; } // Last time any client has connected to server (connection dead or alive)
-        
         public bool Disabled { get; set; }
 
+        // in sharding context - orchestrator node tag
```
Comment (ayende) [inline src/Raven.Client/Documents/Subscriptions/SubscriptionState.cs:28]: Add comment on the normal state as well

File: src/Raven.Server/Documents/Sharding/Subscriptions/ShardedDocumentsDatabaseSubscriptionProcessor.cs [hunk 3]
```diff
@@ -44,4 +59,32 @@ protected override bool ShouldSend(Document item, out string reason, out Excepti
...
+        item.Document.ChangeVector = context.GetChangeVector(cv.Version, cv.Order.RemoveId(_sharding.DatabaseId, context));
+        return true;
+    }
+
+    public HashSet<string> Skipped;
+
+    public override async Task<long> RecordBatch(string lastChangeVectorSentInThisBatch)
+    {
+        var result = await SubscriptionConnectionsState.RecordBatchDocuments(BatchItems, ItemsToRemoveFromResend, lastChangeVectorSentInThisBatch);
+        Skipped = result.Skipped as HashSet<string>;
```
Comment (ayende) [inline src/Raven.Server/Documents/Sharding/Subscriptions/ShardedDocumentsDatabaseSubscriptionProcessor.cs:86]: That s strange code here

File: src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs [hunk 4]
```diff
@@ -254,57 +267,29 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, _buffer))
...
-                                {
-                                    includeCommand.IncludeDocumentsCommand?.Gather(result.Doc);
-                                    includeCommand.IncludeCountersCommand?.Fill(result.Doc);
-                                    includeCommand.IncludeTimeSeriesCommand?.Fill(result.Doc);
-
-                                    writer.WriteDocument(context, result.Doc, metadataOnly: false);
-                                }
-
-                                writer.WriteEndObject();
+                                CurrentBatch.Add(result);
```
Comment (ayende) [inline src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs:270]: what limits the size of a batch?

File: src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsStateForShard.cs [hunk 5]
```diff
@@ -47,10 +49,28 @@ protected override void ValidateTakeOver(SubscriptionConnection currentConnectio
...
+    protected override Task<(long Index, object Skipped)> RecordBatchInternal(RecordBatchSubscriptionDocumentsCommand command)
     {
         command.DatabaseName = ShardedDocumentDatabase.ShardedDatabaseName;
         command.ShardName = ShardedDocumentDatabase.Name;
+        command.ActiveBatchesFromSender = GetActiveBatches();
         return base.RecordBatchInternal(command);
     }
+
+    
+    public bool HasDocumentFormResendForShard()
```
Comment (ayende) [inline src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsStateForShard.cs]: What does this mean? What is this used for?

**Comment by ayende** [comment]
* Revisions subscription - can block for now.
* Not sure that I follow about failover & resharding story, what is the problem here?

File: src/Raven.Client/Documents/Operations/OngoingTasks/GetTaskInfoResult.cs (hunk 1 above)
Comment (karmeli87) [reply src/Raven.Client/Documents/Operations/OngoingTasks/GetTaskInfoResult.cs:98]: we have this in the base class

File: src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs (hunk 4 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs:270]: limited by memory, 4MB for 32bits and 32MB for 64bit
limit by batch size defined at `SubscriptionWorkerOptions.MaxDocsPerBatch`, default 4096 items

File: src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsStateForShard.cs (hunk 5 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsStateForShard.cs]: This is for the single shard connection to check if he can pick anything from the resend queue or go idle and wait for new documents.
used by `FoundAboutMoreDocs/WaitForChangedDocsAsync`

**Comment by karmeli87** [comment]
failover & re-sharding is not working (but failover & sharding is working)

**Comment by ayende** [comment]
What does "failover & re-sharding " mean?

**Comment by karmeli87** [comment]
If I start killing nodes and bringing them back up while rehsarding is in progress - it seems not to work.
I have a failing test which I currently working on.

File: src/Raven.Client/Documents/Subscriptions/SubscriptionState.cs [hunk 6]
```diff
@@ -17,17 +15,20 @@ public class SubscriptionState : IDatabaseTask, IDatabaseTaskStatus
...
+        public string NodeTag { get; set; }
+        
+        public SubscriptionShardingState SubscriptionShardingState { get; set; }
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Subscriptions/SubscriptionState.cs]: rename property to SubscriptionShardingState  => ShardingState

File: src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsStateForShard.cs (hunk 5 above)
Comment (ppekrol) [inline src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsStateForShard.cs]: Form? From? For?

File: src/Raven.Server/Documents/TcpHandlers/SubscriptionConnectionForShard.cs [hunk 7]
```diff
@@ -36,11 +48,67 @@ protected override DynamicJsonValue AcceptMessage()
...
+
+        return false;
+    }
+
     public SubscriptionConnectionsState GetSubscriptionConnectionStateForShard()
     {
         var subscriptions = TcpConnection.DocumentDatabase.SubscriptionStorage.Subscriptions;
         var state = subscriptions.GetOrAdd(SubscriptionId, subId => new SubscriptionConnectionsStateForShard(DatabaseName, subId, TcpConnection.DocumentDatabase.SubscriptionStorage));
         _subscriptionConnectionsState = state;
+        _state = state as SubscriptionConnectionsStateForShard;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/TcpHandlers/SubscriptionConnectionForShard.cs]: are we expecting here null? maybe avoid safe cast or add a Debug.Assert at least?

File: src/Raven.Server/Utils/ChangeVector.cs [hunk 8]
```diff
@@ -75,6 +75,7 @@ public void Renew(ChangeVector version, ChangeVector order)
     public bool IsSingle => string.IsNullOrEmpty(_order?._changeVector) &&
                             string.IsNullOrEmpty(_version?._changeVector);
 
+    public int NumberOfEntries => _changeVector.ToChangeVector().Length;
```
Comment (ppekrol) [inline src/Raven.Server/Utils/ChangeVector.cs]: throw if _changeVector is null?

File: src/Sparrow/Json/JsonDeserializationBase.cs [hunk 9]
```diff
@@ -550,7 +584,7 @@ private static T CreateFromBlittableJson<T>(BlittableJsonReaderObject json, stri
             return dic;
         }
 
-        private static Dictionary<string, List<T>> ToDictionaryOfList<T>(BlittableJsonReaderObject json, string name, JsonDeserializationStringDictionaryAttribute jsonDeserializationDictionaryAttribute, Func<BlittableJsonReaderObject, T> converter)
+        /*private static Dictionary<string, List<T>> ToDictionaryOfList<T>(BlittableJsonReaderObject json, string name, JsonDeserializationStringDictionaryAttribute jsonDeserializationDictionaryAttribute, Func<BlittableJsonReaderObject, T> converter)
```
Comment (ppekrol) [inline src/Sparrow/Json/JsonDeserializationBase.cs]: can be removed?

File: src/Raven.Server/Utils/ChangeVector.cs (hunk 8 above)
Comment (karmeli87) [reply src/Raven.Server/Utils/ChangeVector.cs]: not used, removed.

File: src/Sparrow/Json/JsonDeserializationBase.cs (hunk 9 above)
Comment (karmeli87) [reply src/Sparrow/Json/JsonDeserializationBase.cs]: removed

File: src/Raven.Server/Documents/TcpHandlers/SubscriptionConnectionForShard.cs (hunk 7 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/TcpHandlers/SubscriptionConnectionForShard.cs]: done

File: src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsStateForShard.cs (hunk 5 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsStateForShard.cs]: renamed

File: src/Raven.Client/Documents/Subscriptions/SubscriptionState.cs (hunk 6 above)
Comment (karmeli87) [reply src/Raven.Client/Documents/Subscriptions/SubscriptionState.cs]: done

**Pull request merged**